core/drivers/crypto/hisilicon/hisi_qm.c

```diff
@@ -143,49 +148,108 @@ static void qm_db(struct hisi_qm *qm, uint16_t qn, uint8_t cmd, uint16_t index,
+static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
 {
-	uint32_t val = 0;
-
-	/* Return 0 mailbox ready, HISI_QM_DRVCRYPT_ETMOUT hardware timeout */
-	if (IO_READ32_POLL_TIMEOUT(qm->io_base + QM_MAILBOX_BASE, val,
-				   !(val & QM_MB_BUSY_BIT), POLL_PERIOD,
-				   POLL_TIMEOUT)) {
-		return HISI_QM_DRVCRYPT_ETMOUT;
-	}
+	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
 
-	return HISI_QM_DRVCRYPT_NO_ERR;
+	write_64bit_pair(dst, mb->x[0], mb->x[1]);
 }
 
-static void qm_mb_write(struct hisi_qm *qm, struct qm_mailbox *mb)
+static void qm_mb_read(struct hisi_qm *qm, struct qm_mailbox *mb)
 {
-	vaddr_t dst = qm->io_base + QM_MAILBOX_BASE;
+	vaddr_t mb_base = qm->io_base + QM_MAILBOX_BASE;
 
-	write_64bit_pair(dst, mb->x[0], mb->x[1]);
+	write_64bit_pair((vaddr_t)mb, mb_base, mb_base + U64_OFFSET);
 }
 
-static enum hisi_drv_status qm_mb(struct hisi_qm *qm, uint8_t cmd,
-				  vaddr_t dma_addr, uint16_t qn, uint8_t op)
+static enum hisi_drv_status qm_wait_mb_ready(struct hisi_qm *qm)
 {
 	struct qm_mailbox mb = { };
+	uint32_t timeout = 0;
+
+	timeout = timeout_init_us(QM_MB_WAIT_PERIOD * QM_MB_WAIT_READY_CNT);
+	while (!timeout_elapsed(timeout)) {
+		/* 128 bits should be read from hardware at one time*/
+		qm_mb_read(qm, &mb);
+		if (!(mb.w0 & QM_MB_BUSY_BIT))
+			return HISI_QM_DRVCRYPT_NO_ERR;
+	}
 
-	mb.w0 = cmd | SHIFT_U32(op, QM_MB_OP_SHIFT) |
-		BIT32(QM_MB_BUSY_SHIFT);
-	mb.queue = qn;
-	reg_pair_from_64(dma_addr, &mb.base_h, &mb.base_l);
-	mb.token = 0;
+	EMSG("QM mailbox is busy to start!");
 
-	if (qm_wait_mb_ready(qm)) {
-		EMSG("QM mailbox is busy");
-		return HISI_QM_DRVCRYPT_EBUSY;
+	return HISI_QM_DRVCRYPT_EBUSY;
+}
+
+static enum hisi_drv_status qm_wait_mb_finish(struct hisi_qm *qm,
+					      struct qm_mailbox *mb)
+{
+	uint32_t timeout = 0;
+
+	timeout = timeout_init_us(QM_MB_WAIT_PERIOD * QM_MB_WAIT_MAX_CNT);
+	while (!timeout_elapsed(timeout)) {
+		qm_mb_read(qm, mb);
+		if (!(mb->w0 & QM_MB_BUSY_BIT)) {
+			if (mb->w0 & QM_MB_STATUS_MASK) {
+				EMSG("QM mailbox operation failed!");
+				return HISI_QM_DRVCRYPT_EIO;
+			} else {
+				return HISI_QM_DRVCRYPT_NO_ERR;
+			}
+		}
 	}
 
-	qm_mb_write(qm, &mb);
+	return HISI_QM_DRVCRYPT_ETMOUT;
+}
+
+static void qm_mb_pre_init(struct qm_mailbox *mb, uint8_t cmd, uint64_t base,
+			   uint16_t qnum, uint8_t op)
+{
+	mb->w0 = (cmd | (op ? 0x1 << QM_MB_OP_SHIFT : 0) |
+		  QM_MB_BUSY_BIT);
+	mb->queue = qnum;
+	reg_pair_from_64(base, &mb->base_h, &mb->base_l);
+	mb->token = 0;
+}
 
-	if (qm_wait_mb_ready(qm)) {
-		EMSG("QM mailbox operation timeout");
+static enum hisi_drv_status qm_mb_nolock(struct hisi_qm *qm,
+					 struct qm_mailbox *mb)
+{
+	if (qm_wait_mb_ready(qm))
 		return HISI_QM_DRVCRYPT_EBUSY;
-	}
+
+	qm_mb_write(qm, mb);
+
+	return qm_wait_mb_finish(qm, mb);
+}
+
+static enum hisi_drv_status hisi_qm_mb_write(struct hisi_qm *qm, uint8_t cmd,
+					     uintptr_t dma_addr, uint16_t qnum)
+{
+	struct qm_mailbox mb;
```

Comment:
Please initialize all local variables, this is required by the coding guidelines.